### PR TITLE
Reverted the "kind" field to the previous value.

### PR DESCRIPTION
**Pull request by KalitaAlexey**
Added a new field named "crate_types".
Fixes https://github.com/rust-lang/cargo/issues/3654
After this PR `cargo metadata` would output:
```json
{
    "packages": [
        {
            "targets": [
                {
                    "kind": ["lib"],
                    "crate_types": ["lib"]
                },
                {
                    "kind": ["example"],
                    "crate_types": ["staticlib"]
                }
            ]
        }
    ]
}
```

I have added tests.

### CODE REVIEW

**Comment by alexcrichton** [comment]
I believe previously `kind` for something like `crate-type = ["rlib", "dylib"]` would be `["rlib", "dylib"]`, right? Could we preserve that part (odd though it may be) as well?

**Comment by KalitaAlexey** [comment]
@alexcrichton,
That wasn't changed.

**Comment by KalitaAlexey** [comment]
I will fix it.

**Comment by alexcrichton** [comment]
@KalitaAlexey yes, it was changed. If you make a small example library with two crate-types, metadata prints this

```json
"kind" : [
   "rlib",
   "dylib"
]
```

**Comment by KalitaAlexey** [comment]
@alexcrichton,
I have tests, which tell you are not right, but I will check it.

**Comment by KalitaAlexey** [comment]
build-auth.rs
http_auth_offered test failed:
```
 11 - |attempted to find username/password via [..]`credential.helper`, but [..]|
    + |attempted to find username/password via git's `credential.helper` support, but failed|
```
Windows 10, what am I to do?

**Comment by KalitaAlexey** [comment]
Hm.
I did:
```
git pull upstream master
cargo test
```
I got:
```
failures:

---- finds_author_email stdout ----
        running `D:\Projects\Rust\cargo\target\debug\cargo new foo`
thread 'finds_author_email' panicked at 'assertion failed: contents.contains(r#"authors = ["bar <baz>"]"#)', tests\new.rs:314
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- finds_author_user stdout ----
        running `D:\Projects\Rust\cargo\target\debug\cargo new foo`
thread 'finds_author_user' panicked at 'assertion failed: contents.contains(r#"authors = ["foo"]"#)', tests\new.rs:246

---- finds_author_user_escaped stdout ----
        running `D:\Projects\Rust\cargo\target\debug\cargo new foo`
thread 'finds_author_user_escaped' panicked at 'assertion failed: contents.contains(r#"authors = ["foo \"bar\""]"#)', tests\new.rs:261

---- finds_author_username stdout ----
        running `D:\Projects\Rust\cargo\target\debug\cargo new foo`
thread 'finds_author_username' panicked at 'assertion failed: contents.contains(r#"authors = ["foo"]"#)', tests\new.rs:278

---- finds_git_author stdout ----
        running `D:\Projects\Rust\cargo\target\debug\cargo new foo`
thread 'finds_git_author' panicked at 'assertion failed: contents.contains(r#"authors = ["gitfoo"]"#)', tests\new.rs:362


failures:
    finds_author_email
    finds_author_user
    finds_author_user_escaped
    finds_author_username
    finds_git_author

test result: FAILED. 22 passed; 5 failed; 0 ignored; 0 measured
```

File: tests/metadata.rs [hunk 1]
```diff
@@ -191,6 +203,127 @@ fn cargo_metadata_with_deps_and_version() {
...
+                "targets": [
+                    {
+                        "kind": [ "lib" ],
+                        "crate_types": [ "lib" ],
+                        "name": "foo",
+                        "src_path": "[..][/]foo[/]src[/]lib.rs"
+                    },
+                    {
+                        "kind": [ "example" ],
+                        "crate_types": [ "rlib", "dylib" ],
```
Comment (KalitaAlexey) [inline tests/metadata.rs]: @alexcrichton,
The test checks what you asked.

**Comment by alexcrichton** [comment]
As I [mentioned before](https://github.com/rust-lang/cargo/pull/3668#issuecomment-278233941), we'll need to preserve the existing stable semantics of `cargo metadata` (the purpose of this change). To that end the `kind` entry may need to be an array of crate types actually for libraries. Does that make sense?

**Comment by KalitaAlexey** [comment]
@alexcrichton,
I am sorry, but I can't figure out what you are telling me.

**Comment by alexcrichton** [comment]
Currently this PR will emit at most one entry in the array of the `kind` field, but contrast that with:

```
cargo +stable new foo
cd foo
cat >> Cargo.toml <<-EOF
[lib]
crate-type = ["lib", "dylib"]
EOF
cargo +stable metadata | json_pp | grep -A 3 kind
```

That prints:

```
"kind" : [
   "lib",
   "dylib"
],
```

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by ehuss** [comment]
I'm curious, will this get merged to the beta channel/rustc-1.16?  I'm not familiar with how rust/cargo releases and branches are handled.

**Comment by alexcrichton** [comment]
@ehuss oh is this bug on beta? I thought it was only on nightly?

**Comment by ehuss** [comment]
The original commit that added support for examples as libraries (#3556) was added to master around Jan 20.  I think the rust-1.16.0 branch was made around Feb 1.  This fix was committed to master on Feb 11, after the rust-1.16.0 branch was made.  I assume beta is being pulled from the rust-1.16.0 branch.

**Comment by alexcrichton** [comment]
@ehuss ah yes I'll backport, thanks for the ping!

**Comment by alexcrichton** [comment]
@ehuss backported in https://github.com/rust-lang/cargo/pull/3826

**Comment by KalitaAlexey** [comment]
@alexcrichton,
Will 1.16 be published without the bug?

**Comment by alexcrichton** [comment]
Yes

**Comment by brson** [comment]
This is not actually in the 1.16 release as the backport occured after the build.

**Pull request merged**